### PR TITLE
refactoring: extract write cache from DIMAP, implement in ENVI writer…

### PR DESCRIPTION
**Pull request by TomBlock**
…, added stream synchronisation in ENVI writer

### CODE REVIEW

File: snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java [hunk 1]
```diff
@@ -26,25 +28,74 @@
...
      * Construct a new instance of a product writer for the given ENVI product writer plug-in.
      *
      * @param writerPlugIn the given ENVI product writer plug-in, must not be <code>null</code>
      */
     public EnviProductWriter(ProductWriterPlugIn writerPlugIn) {
         super(writerPlugIn);
+
+        final Preferences preferences = Config.instance().preferences();
+        useCache = preferences.getBoolean(SYSPROP_USE_CACHE, true);
+        writeCache = new WriteCache();
```
Comment (SabineEmbacher) [inline snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java:58]: Why is a WriteCache instantiated even if useCache is false?

File: snap-core/src/main/java/org/esa/snap/core/dataio/dimap/DimapProductWriter.java [hunk 2]
```diff
@@ -191,17 +251,17 @@ public void writeBandRasterData(Band sourceBand,
...
-            // @todo 1 tb/tb progress monitoring 2020-07-03
             final VariableCache variableCache = writeCache.get(sourceBand);
             final boolean canWrite = variableCache.update(sourceOffsetX, sourceOffsetY, sourceWidth, sourceHeight, sourceBuffer);
             if (canWrite) {
-                synchronized (outputStream){
+                synchronized (outputStream) {
                     variableCache.writeCompletedBlocks(outputStream);
                 }
             }
+            pm.worked(1);
```
Comment (SabineEmbacher) [inline snap-core/src/main/java/org/esa/snap/core/dataio/dimap/DimapProductWriter.java:264]: Not valid without pm.beginTask(...).
Also it should be wrapped with a try/finally block. Place pm.done() in the finally block.

File: snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java [hunk 3]
```diff
@@ -114,17 +165,31 @@ public void writeBandRasterData(Band sourceBand,
...
+        final boolean fullRaster = isFullRaster(sourceWidth, sourceHeight, sourceBandWidth, sourceBandHeight);
+        if (useCache && (!fullRaster)) {
+            final VariableCache variableCache = writeCache.get(sourceBand);
+            final boolean canWrite = variableCache.update(sourceOffsetX, sourceOffsetY, sourceWidth, sourceHeight, sourceBuffer);
+            if (canWrite) {
+                synchronized (outputStream) {
+                    variableCache.writeCompletedBlocks(outputStream);
+                }
             }
             pm.worked(1);
```
Comment (SabineEmbacher) [inline snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java:177]: Not valid without pm.beginTask(...).
Also it should be wrapped with a try/finally block. Place pm.done() in the finally block.

File: snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java (hunk 1 above)
Comment (TomBlock) [reply snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java:58]: resolved - create conditional

File: snap-core/src/main/java/org/esa/snap/core/dataio/dimap/DimapProductWriter.java (hunk 2 above)
Comment (TomBlock) [reply snap-core/src/main/java/org/esa/snap/core/dataio/dimap/DimapProductWriter.java:264]: resolved - corrected progressbar handling

File: snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java (hunk 3 above)
Comment (TomBlock) [reply snap-envi-reader/src/main/java/org/esa/snap/dataio/envi/EnviProductWriter.java:177]: resolved - corrected progressbar handling

**Pull request merged**